flash/data/data_pipeline.py

```diff
@@ -0,0 +1,612 @@
+
+    1. Generate an ``AutoDataset`` from ``load_data`` and ``load_sample``.
+
+        Example::
+            class AutoDataset
+                def __init__(self, ..., data, ...):
+                    self.preprocessed_data: Iterable = Preprocess.load_data(data)
+
+                def __getitem__(self, index):
+                    return Preprocess.load_sample(self.preprocessed_data[index])
+
+                def __len__(self):
+                    return len(self.preprocessed_data)
+
+    2. Create a ``worker_collate_fn`` which is injected directly into the ``DataLoader``
+       and a ``device_collate_fn`` injected after ``LightningModule.transfer_batch_to_device`` hook.
+
+        Objects description:
+
+        _Sequential:
+            ┌────────────────────────────────────┐
+            │  per_sample_pre_tensor_transform   │
+            │                |                   │
+            │  per_sample_to_tensor_transform    │
+            │                |                   │
+            │  per_sample_post_tensor_transform  │
+            └────────────────────────────────────┘
+
+        _PreProcessor:
+
+            The ``_PreProcessor`` performs ``per_sample_transform``, ``collate``, ``per_batch_transform`` as follow:
+
+            ``per_batch_transform`` and ``per_sample_transform_on_device`` are mutually exclusive
+
+            def forward(self, samples: Sequence[Any]):
+                    samples = [self.per_sample_transform(sample) for sample in samples]
+                    samples = type(samples)(samples)
+                    samples = self.collate_fn(samples)
+                samples = self.per_batch_transform(samples)
+                return samples
+
+            ``_PreProcessor`` in worker:
+
+                * per_sample_transform: _Sequential(
+                    per_sample_pre_tensor_transform, per_sample_to_tensor_transform, per_sample_post_tensor_transform)
+
+                * collate: Set to ``do_nothing`` is ``per_sample_transform_on_device`` is implemented
+                    and not ``per_batch_transform``
+
+                * per_batch_transform
+
+            ``_PreProcessor`` on device:
+
+                * per_sample_transform_on_device
+
+                * collate: Set to ``do_nothing`` is ``per_batch_transform`` is implemented
+                    and not ``per_sample_transform_on_device``
+
+                * per_batch_transform_on_device
+
+
+        General flow:
+                                             load_sample
+                                                 │
+                                   per_sample_pre_tensor_transform
+                                                 │
+                                    per_sample_to_tensor_transform
+                                                 │
+                                  per_sample_post_tensor_transform
+                                                 │
+                                ┌────────────────┴───────────────────┐
```

Comment:
do I miss something or hos this branching is happening?

Reply:
Yes. It does depending on which hooks you override.